clients/application-kafka/src/main/java/org/eclipse/hono/application/client/kafka/impl/KafkaBasedCommandSender.java

```diff
@@ -157,4 +264,123 @@ public KafkaBasedCommandSender(
 
         return props;
     }
+
+    private Handler<DownstreamMessage<KafkaMessageContext>> getCommandResponseHandler(final String tenantId,
+            final Span span) {
+        return message -> {
+            if (message.getCorrelationId() != null) {
+                // if the correlation id of the received response matches with that of the command
+                Optional.ofNullable(pendingCommandResponses.get(tenantId))
+                        .map(ids -> ids.remove(message.getCorrelationId()))
+                        .ifPresent(expiringPromise -> {
```

Comment:
How about using `ifPresentOrElse` here instead and adding a `LOGGER.debug(` line about a command response message having been received that couldn't be associated with a sent command (potentially because waiting for the response has timed out).